File: elixir/scrabble-score/scrabble.exs
defmodule Scrabble do

    @points %{
        A: 1, E: 1, I: 1, O: 1, U: 1, L: 1, N: 1, R: 1, S: 1, T: 1,
        D: 2, G: 2,
        B: 3, C: 3, M: 3, P: 3,
        F: 4, H: 4, V: 4, W: 4, Y: 4,
        K: 5,
        J: 8, X: 8,
        Q: 10, Z: 10
    }

  @doc """
  Calculate the scrabble score for the word.
  """
  @spec score(String.t) :: non_neg_integer
  def score(word) do
    word
        |> to_letters()
        |> get_score()
  end

  defp to_letters(word) do
    word
        |> String.upcase()
        |> String.replace(~r/[^A-Z]/, "")
        |> String.graphemes()
  end

  defp get_score(word) do
    word
        |> Enum.reduce(0, fn(letter, score) -> get_points(letter) + score end)
  end

  defp get_points(letter), do: @points[String.to_atom(letter)]

end

File: java/scrabble-score/src/main/java/Scrabble.java
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

class Scrabble {

    private static Map<Character, Integer> letterScore;
    private static final Map<Character, Integer> LETTER_SCORE;
    static {
        letterScore = new HashMap<>();
        initLetterScores("AEIOULNRST", 1);
        initLetterScores("DG", 2);
        initLetterScores("BCMP", 3);
        initLetterScores("FHVWY", 4);
        initLetterScores("K", 5);
        initLetterScores("JX", 8);
        initLetterScores("QZ", 10);
        LETTER_SCORE = Collections.unmodifiableMap(letterScore);
    }

    private static void initLetterScores(String letters, int points) {
        for (char letter : letters.toCharArray()) {
            letterScore.put(letter,
                    points);
        }
    }

    private String word;

    Scrabble(String word) {

        this.word = word.toUpperCase();
    }

    int getScore() {
        int score = 0;

        for (char letter : word
                .toCharArray()) {

            score += LETTER_SCORE.get(letter);
        }

        return score;
    }
}
